servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleZipper.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.Publisher.from;
+import static java.util.Comparator.comparingInt;
+import static java.util.function.Function.identity;
+
+final class SingleZipper {
+    private static final Comparator<? super ZipArg> ZIP_ARG_COMPARATOR = comparingInt(o -> o.index);
+
+    private SingleZipper() {
+    }
+
+    @SuppressWarnings("unchecked")
+    static <T1, T2, R> Single<R> zip(Single<? extends T1> s1, Single<? extends T2> s2,
+                                     BiFunction<? super T1, ? super T2, ? extends R> zipper) {
+        return from(s1.map(v -> new ZipArg(0, v)), s2.map(v -> new ZipArg(1, v)))
+                .flatMapMergeSingle(identity(), 2)
+                .collect(ZipResult2::new, (result, zipArg) -> {
```

Comment:
Can we replace the conditionals with a direct array assignment?
example:
```
static <T1, T2, R> Single<R> zip(Single<? extends T1> s1, Single<? extends T2> s2,
                                 BiFunction<? super T1, ? super T2, ? extends R> zipper) {
    return from(s1.map(v -> new ZipArg(0, v)), s2.map(v -> new ZipArg(1, v)))
            .flatMapMergeSingle(identity(), 2)
            .collect(ZipResult2::new, (result, zipArg) -> {
                result.args[zipArg.index] = zipArg;
                return result;
            }).map(result -> zipper.apply((T1) result.args[0].value, (T2) result.args[1].value));
}
```

Reply:
> Can we replace the conditionals with a direct array assignment?

+1. good suggestion!

> nit: you could probably access the arr elements in reverse order and store them in local fields before calling the zipper, to minimize array bound checks.

I'm not sure I follow. can you drop another comment after I submit the updates if I don't address this?

Reply:
What I meant is this:

```
final Object c = result.args[2].value;
final Object b = result.args[1].value;
final Object a = result.args[0].value;

zipper.apply((T1) a, (T2) b, (T3) c))
```

This should generate code that checks array boundaries only once (1st time) and make the subsequent accesses cheaper (if JIT doesn't handle this case already). This of course, at the cost of extra code, which may confuse readers, so not sure its worth it here.

Reply:
neat. lets proceed without this for now and we can submit a follow-up to use this optimization if it proves necessary/beneficial.